core/coreapi/coreapi.go

```diff
@@ -16,19 +16,26 @@ package coreapi
 import (
 	core "github.com/ipfs/go-ipfs/core"
 	coreiface "github.com/ipfs/go-ipfs/core/coreapi/interface"
-
 	logging "gx/ipfs/QmZChCsSt8DctjceaL56Eibc29CVQq4dGKRXC5JRZ6Ppae/go-log"
+	ipld "gx/ipfs/QmdDXJs4axxefSPgK6Y1QhpJWKuDPnGJiqgq4uncb4rFHL/go-ipld-format"
 )
 
 var log = logging.Logger("core/coreapi")
 
 type CoreAPI struct {
 	node *core.IpfsNode
+	dag  ipld.DAGService
```

Comment:
This is going to be a footgun. Now the CoreAPI needs to make sure to use *it's* dag instead of the global one.

Maybe https://github.com/ipfs/go-ipfs/issues/4009? This keeps on haunting us and I've yet to see a better proposal.